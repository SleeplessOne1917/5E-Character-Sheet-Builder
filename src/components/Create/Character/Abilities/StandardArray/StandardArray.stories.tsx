import { ComponentMeta, ComponentStory } from '@storybook/react';
import { AbilityScoresState } from '../../../../../redux/features/abilityScores';
import MockStore, { mockAbilities } from '../MockAbilitiesStore';
import StandardArray, { StandardArrayProps } from './StandardArray';

export default {
	title: 'Components/CharacterCreation/Abilities/StandardArray',
	component: StandardArray,
	args: {
		abilities: mockAbilities
	}
} as ComponentMeta<typeof StandardArray>;

const Template: ComponentStory<typeof StandardArray> = (
	args: StandardArrayProps
) => <StandardArray {...args} />;

const defaultOverride: AbilityScoresState = {
	str: { highest: 20, abilityImprovement: 0 },
	dex: { highest: 20, abilityImprovement: 0 },
	con: { highest: 20, abilityImprovement: 0 },
	int: { highest: 20, abilityImprovement: 0 },
	cha: { highest: 20, abilityImprovement: 0 },
	wis: { highest: 20, abilityImprovement: 0 }
};

export const NoneSelected = Template.bind({});
NoneSelected.decorators = [
	story => <MockStore overrideValues={defaultOverride}>{story()}</MockStore>
];

export const HalfSelected = Template.bind({});
HalfSelected.decorators = [
	story => (
		<MockStore
			overrideValues={{
				...defaultOverride,
				str: { base: 8, highest: 20, abilityImprovement: 0 },
				int: { base: 15, highest: 20, abilityImprovement: 0 },
				con: { base: 14, highest: 20, abilityImprovement: 0 }
			}}
		>
			{story()}
		</MockStore>
	)
];

export const AllSelected = Template.bind({});
AllSelected.decorators = [
	story => (
		<MockStore
			overrideValues={{
				str: { base: 8, highest: 20, abilityImprovement: 0 },
				dex: { base: 10, highest: 20, abilityImprovement: 0 },
				con: { base: 12, highest: 20, abilityImprovement: 0 },
				int: { base: 13, highest: 20, abilityImprovement: 0 },
				cha: { base: 14, highest: 20, abilityImprovement: 0 },
				wis: { base: 15, highest: 20, abilityImprovement: 0 }
			}}
		>
			{story()}
		</MockStore>
	)
];
