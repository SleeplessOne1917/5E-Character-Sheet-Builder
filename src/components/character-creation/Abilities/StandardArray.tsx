import { ChangeEvent, useCallback, useState } from 'react';
import { useAppDispatch, useAppSelector } from '../../../hooks/reduxHooks';

import { AbilityItem } from '../../../types/srd';
import AbilityScores from '../../../types/abilityScores';
import classes from './StandardArray.module.css';
import { getTotalScore } from '../../../services/abilityScoreService';
import { updateBase } from '../../../redux/features/abilityScores';
import useGetAbilityScore from '../../../hooks/useGetAbilityScore';

const arrayValues = [8, 10, 12, 13, 14, 15];

type StandardArrayProps = {
	abilities: AbilityItem[];
};

const StandardArray = ({ abilities }: StandardArrayProps): JSX.Element => {
	const dispatch = useAppDispatch();
	const getAbilityScore = useGetAbilityScore();
	const abilityScores = useAppSelector(
		state => state.editingCharacter.abilityScores
	);
	const handleValueSelect = useCallback(
		(event: ChangeEvent<HTMLSelectElement>, abilityIndex: AbilityScores) => {
			const newValue = event.target.value;

			if (newValue === 'blank') {
				dispatch(updateBase({ value: null, abilityIndex }));
			} else {
				dispatch(updateBase({ value: parseInt(newValue), abilityIndex }));
			}
		},
		[dispatch]
	);

	return (
		<div className={classes.abilities}>
			{abilities.map(ability => {
				const abilityScore = getAbilityScore(ability.index as AbilityScores);

				return (
					<div key={ability.index} className={classes.ability}>
						<h3>{ability.full_name}</h3>
						<select
							onChange={event =>
								handleValueSelect(event, ability.index as AbilityScores)
							}
							value={abilityScore.base ? abilityScore.base : 'blank'}
						>
							<option value="blank">&mdash;</option>
							{(abilityScore.base ? [abilityScore.base] : [])
								.concat(
									arrayValues.filter(
										value =>
											!Object.values(abilityScores)
												.map(score => score.base)
												.includes(value)
									)
								)
								.sort((a, b) => {
									if (a < b) {
										return -1;
									} else if (a > b) {
										return 1;
									} else {
										return 0;
									}
								})
								.map(value => (
									<option key={value} value={value}>
										{value}
									</option>
								))}
						</select>
						<h4>
							Total:{' '}
							{abilityScore.base ? getTotalScore(abilityScore) : '\u2014'}
						</h4>
					</div>
				);
			})}
		</div>
	);
};

export default StandardArray;
