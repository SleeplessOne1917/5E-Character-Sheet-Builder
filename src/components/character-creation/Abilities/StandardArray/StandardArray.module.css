.abilities {
    display: flex;
    flex-wrap: wrap;
    align-self: center;
    width: 100%;
}

.ability {
    width: 28%;
    color: var(--davy);
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 0 0.5rem;
}

.ability div {
    font-weight: bold;
    margin: 1rem 0;
}

.ability label {
    font-size: 1.2rem;
    font-weight: bold;
    margin: 1rem 0;
}

.ability select {
    width: 50%;
    border: 1px solid var(--davy);
    font-size: 1.5rem;
    font-family: inherit;
    background-color: var(--heaven-light);
}

@media screen and (min-width: 768px) {
    .ability {
        width: 30%;
    }
}

@media screen and (min-width: 1024px) {
    .ability {
        width: 14.5%;
    }
}

@media screen and (min-width: 1280px) {
    .ability {
        width: 14%;
    }
}

@media screen and (min-width: 1984px) {
    .ability {
        width: 15%;
    }
}