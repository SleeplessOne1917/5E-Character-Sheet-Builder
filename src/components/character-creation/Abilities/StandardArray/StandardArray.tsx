import { useAppDispatch, useAppSelector } from '../../../../hooks/reduxHooks';

import { AbilityItem } from '../../../../types/srd';
import AbilityScores from '../../../../types/abilityScores';
import Select from '../../../Select/Select/Select';
import classes from './StandardArray.module.css';
import { getTotalScore } from '../../../../services/abilityScoreService';
import {
	AbilityScore,
	updateBase
} from '../../../../redux/features/abilityScores';
import { useCallback } from 'react';
import useGetAbilityScore from '../../../../hooks/useGetAbilityScore';
import usePreparedSpells from '../../../../hooks/usePreparedSpells';
import { removeSpell } from '../../../../redux/features/spellcasting';
import { removeClassSpell } from '../../../../redux/features/classInfo';

const arrayValues = [8, 10, 12, 13, 14, 15];

export type StandardArrayProps = {
	abilities: AbilityItem[];
};

const StandardArray = ({ abilities }: StandardArrayProps): JSX.Element => {
	const dispatch = useAppDispatch();
	const getAbilityScore = useGetAbilityScore();
	const abilityScores = useAppSelector(
		state => state.editingCharacter.abilityScores
	);
	const spellcastingAbility = useAppSelector(
		state =>
			state.editingCharacter.classInfo.class?.spellcasting?.spellcasting_ability
				.index
	);
	const { getNumberOfSpellsToPrepare, shouldPrepareSpells } =
		usePreparedSpells();
	const classSpells = useAppSelector(
		state => state.editingCharacter.classInfo.spells
	)?.filter(({ level }) => level > 0);

	const handleValueSelect = useCallback(
		(value: string | number, abilityIndex: AbilityScores) => {
			if (value === 'blank') {
				dispatch(updateBase({ value: null, abilityIndex }));
			} else {
				dispatch(updateBase({ value: value as number, abilityIndex }));
			}

			if (
				spellcastingAbility &&
				abilityIndex === spellcastingAbility &&
				shouldPrepareSpells
			) {
				const newPreparedSpellsNumber = getNumberOfSpellsToPrepare({
					abilityScore: {
						...(
							abilityScores as {
								[key: string]: AbilityScore;
							}
						)[abilityIndex],
						base: value === 'blank' ? null : (value as number)
					}
				});

				if (classSpells && newPreparedSpellsNumber < classSpells.length) {
					for (
						let i = 0;
						i < classSpells.length - newPreparedSpellsNumber;
						++i
					) {
						const spellToRemove = classSpells[classSpells.length - (i + 1)];

						dispatch(removeSpell(spellToRemove.index));
						dispatch(removeClassSpell(spellToRemove.index));
					}
				}
			}
		},
		[
			dispatch,
			abilityScores,
			classSpells,
			getNumberOfSpellsToPrepare,
			shouldPrepareSpells,
			spellcastingAbility
		]
	);

	return (
		<div
			className={classes.abilities}
			role="region"
			aria-label="Standard Array"
		>
			{abilities.map(ability => {
				const abilityScore = getAbilityScore(ability.index as AbilityScores);

				return (
					<div key={ability.index} className={classes.ability}>
						<Select
							fontSize="1.5rem"
							testId={ability.index}
							label={ability.full_name}
							id={`${ability.index}-select`}
							onChange={value =>
								handleValueSelect(value, ability.index as AbilityScores)
							}
							value={abilityScore.base ? abilityScore.base : 'blank'}
							options={(
								[{ value: 'blank', label: '\u2014' }] as {
									value: string | number;
									label: string;
								}[]
							).concat(
								(abilityScore.base ? [abilityScore.base] : [])
									.concat(
										arrayValues.filter(
											value =>
												!Object.values(abilityScores)
													.map(score => score.base)
													.includes(value)
										)
									)
									.sort((a, b) => a - b)
									.map(value => ({
										value: value,
										label: `${value}`
									}))
							)}
						/>
						<div>
							Total:{' '}
							{abilityScore.base ? getTotalScore(abilityScore) : '\u2014'}
						</div>
					</div>
				);
			})}
		</div>
	);
};

export default StandardArray;
