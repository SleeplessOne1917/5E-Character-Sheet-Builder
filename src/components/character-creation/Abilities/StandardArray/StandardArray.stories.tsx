import { ComponentMeta, ComponentStory } from '@storybook/react';
import MockStore, { mockAbilities } from '../MockAbilitiesStore';
import StandardArray, { StandardArrayProps } from './StandardArray';

const StandardArraySb = {
	title: 'Components/StandardArray',
	component: StandardArray,
	args: {
		abilities: mockAbilities
	}
} as ComponentMeta<typeof StandardArray>;

export default StandardArraySb;

const Template: ComponentStory<typeof StandardArray> = (
	args: StandardArrayProps
) => <StandardArray {...args} />;

const defaultOverride = {
	str: {},
	dex: {},
	con: {},
	int: {},
	cha: {},
	wis: {}
};

export const NoneSelected = Template.bind({});
NoneSelected.decorators = [
	story => <MockStore overrideValues={defaultOverride}>{story()}</MockStore>
];

export const HalfSelected = Template.bind({});
HalfSelected.decorators = [
	story => (
		<MockStore
			overrideValues={{
				...defaultOverride,
				str: { base: 8 },
				int: { base: 15 },
				con: { base: 14 }
			}}
		>
			{story()}
		</MockStore>
	)
];

export const AllSelected = Template.bind({});
AllSelected.decorators = [
	story => (
		<MockStore
			overrideValues={{
				str: { base: 8 },
				dex: { base: 10 },
				con: { base: 12 },
				int: { base: 13 },
				cha: { base: 14 },
				wis: { base: 15 }
			}}
		>
			{story()}
		</MockStore>
	)
];
