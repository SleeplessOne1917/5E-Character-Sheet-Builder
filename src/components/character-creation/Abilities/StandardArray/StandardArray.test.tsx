import '@testing-library/jest-dom';

import * as stories from './StandardArray.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';

const { HalfSelected, NoneSelected } = composeStories(stories);

const getSelectOptions = (select: HTMLDivElement) => {
	const options = [] as HTMLLIElement[];
	/* eslint-disable testing-library/no-node-access */
	for (let i = 0; i < (select.querySelector('ul')?.children.length ?? 0); ++i) {
		options.push(select.querySelector('ul')?.children.item(i) as HTMLLIElement);
	}
	/* eslint-enable testing-library/no-node-access */

	return options.map(option => option.textContent);
};

it('renders correctly', () => {
	render(<HalfSelected />);

	expect(screen.getByRole('region')).toMatchSnapshot();
});

it('only renders total score when score is selected', async () => {
	render(<NoneSelected />);

	expect(screen.queryByText(/Total: \d/i)).not.toBeInTheDocument();
	const originalNumberOfEmdashTotals =
		screen.getAllByText(/Total: \u2014/).length;
	expect(originalNumberOfEmdashTotals).toBeGreaterThan(0);

	await userEvent.selectOptions(screen.getByTestId('str-list'), '8');

	expect(screen.getByText(/Total: \d/i)).toBeInTheDocument();
	expect(screen.getAllByText(/Total: \u2014/).length).toBe(
		originalNumberOfEmdashTotals - 1
	);
});

it('removes selected options from the options of other dropdowns', async () => {
	render(<NoneSelected />);

	const selected = [];
	await userEvent.selectOptions(screen.getByTestId('str-list'), '8');
	selected.push('8');

	const secondSelect = screen.getByLabelText(/Dexterity/i) as HTMLDivElement;
	for (const s of selected) {
		expect(getSelectOptions(secondSelect)).not.toContain(s);
	}
	await userEvent.selectOptions(screen.getByTestId('dex-list'), '10');
	selected.push('10');

	const thirdSelect = screen.getByLabelText(/Constitution/i) as HTMLDivElement;
	for (const s of selected) {
		expect(getSelectOptions(thirdSelect)).not.toContain(s);
	}
	await userEvent.selectOptions(screen.getByTestId('con-list'), '12');
	selected.push('12');

	const fourthSelect = screen.getByLabelText(/Intelligence/i) as HTMLDivElement;
	for (const s of selected) {
		expect(getSelectOptions(fourthSelect)).not.toContain(s);
	}
	await userEvent.selectOptions(screen.getByTestId('int-list'), '13');
	selected.push('13');

	const fifthSelect = screen.getByLabelText(/Wisdom/i) as HTMLDivElement;
	for (const s of selected) {
		expect(getSelectOptions(fifthSelect)).not.toContain(s);
	}
	await userEvent.selectOptions(screen.getByTestId('wis-list'), '14');
	selected.push('14');

	const sixthSelect = screen.getByLabelText(/Charisma/i) as HTMLDivElement;
	for (const s of selected) {
		expect(getSelectOptions(sixthSelect)).not.toContain(s);
	}
});
