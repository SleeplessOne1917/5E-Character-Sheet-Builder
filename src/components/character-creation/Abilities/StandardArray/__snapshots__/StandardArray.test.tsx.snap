// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  aria-label="Standard Array"
  class="abilities"
  role="region"
>
  <div
    class="ability"
  >
    <label
      id="str-select-label"
    >
      Strength
    </label>
    <div
      aria-labelledby="str-select-label"
      class="container"
      data-testid="str"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="button"
        data-testid="str-button"
        type="button"
      >
        8
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <ul
        aria-activedescendant="8"
        class="list"
        data-testid="str-list"
        role="listbox"
        style="display: none;"
        tabindex="-1"
      >
        <li
          aria-selected="false"
          class="item"
          data-testid="str-data-blank"
          role="option"
          tabindex="0"
        >
          —
        </li>
        <li
          aria-selected="true"
          class="item selected"
          data-testid="str-data-8"
          role="option"
          tabindex="0"
        >
          8
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="str-data-10"
          role="option"
          tabindex="0"
        >
          10
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="str-data-12"
          role="option"
          tabindex="0"
        >
          12
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="str-data-13"
          role="option"
          tabindex="0"
        >
          13
        </li>
      </ul>
    </div>
    <div>
      Total:
       
      8
    </div>
  </div>
  <div
    class="ability"
  >
    <label
      id="dex-select-label"
    >
      Dexterity
    </label>
    <div
      aria-labelledby="dex-select-label"
      class="container"
      data-testid="dex"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="button"
        data-testid="dex-button"
        type="button"
      >
        —
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <ul
        aria-activedescendant="—"
        class="list"
        data-testid="dex-list"
        role="listbox"
        style="display: none;"
        tabindex="-1"
      >
        <li
          aria-selected="true"
          class="item selected"
          data-testid="dex-data-blank"
          role="option"
          tabindex="0"
        >
          —
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="dex-data-10"
          role="option"
          tabindex="0"
        >
          10
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="dex-data-12"
          role="option"
          tabindex="0"
        >
          12
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="dex-data-13"
          role="option"
          tabindex="0"
        >
          13
        </li>
      </ul>
    </div>
    <div>
      Total:
       
      —
    </div>
  </div>
  <div
    class="ability"
  >
    <label
      id="con-select-label"
    >
      Constitution
    </label>
    <div
      aria-labelledby="con-select-label"
      class="container"
      data-testid="con"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="button"
        data-testid="con-button"
        type="button"
      >
        14
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <ul
        aria-activedescendant="14"
        class="list"
        data-testid="con-list"
        role="listbox"
        style="display: none;"
        tabindex="-1"
      >
        <li
          aria-selected="false"
          class="item"
          data-testid="con-data-blank"
          role="option"
          tabindex="0"
        >
          —
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="con-data-10"
          role="option"
          tabindex="0"
        >
          10
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="con-data-12"
          role="option"
          tabindex="0"
        >
          12
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="con-data-13"
          role="option"
          tabindex="0"
        >
          13
        </li>
        <li
          aria-selected="true"
          class="item selected"
          data-testid="con-data-14"
          role="option"
          tabindex="0"
        >
          14
        </li>
      </ul>
    </div>
    <div>
      Total:
       
      14
    </div>
  </div>
  <div
    class="ability"
  >
    <label
      id="int-select-label"
    >
      Intelligence
    </label>
    <div
      aria-labelledby="int-select-label"
      class="container"
      data-testid="int"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="button"
        data-testid="int-button"
        type="button"
      >
        15
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <ul
        aria-activedescendant="15"
        class="list"
        data-testid="int-list"
        role="listbox"
        style="display: none;"
        tabindex="-1"
      >
        <li
          aria-selected="false"
          class="item"
          data-testid="int-data-blank"
          role="option"
          tabindex="0"
        >
          —
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="int-data-10"
          role="option"
          tabindex="0"
        >
          10
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="int-data-12"
          role="option"
          tabindex="0"
        >
          12
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="int-data-13"
          role="option"
          tabindex="0"
        >
          13
        </li>
        <li
          aria-selected="true"
          class="item selected"
          data-testid="int-data-15"
          role="option"
          tabindex="0"
        >
          15
        </li>
      </ul>
    </div>
    <div>
      Total:
       
      15
    </div>
  </div>
  <div
    class="ability"
  >
    <label
      id="wis-select-label"
    >
      Wisdom
    </label>
    <div
      aria-labelledby="wis-select-label"
      class="container"
      data-testid="wis"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="button"
        data-testid="wis-button"
        type="button"
      >
        —
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <ul
        aria-activedescendant="—"
        class="list"
        data-testid="wis-list"
        role="listbox"
        style="display: none;"
        tabindex="-1"
      >
        <li
          aria-selected="true"
          class="item selected"
          data-testid="wis-data-blank"
          role="option"
          tabindex="0"
        >
          —
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="wis-data-10"
          role="option"
          tabindex="0"
        >
          10
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="wis-data-12"
          role="option"
          tabindex="0"
        >
          12
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="wis-data-13"
          role="option"
          tabindex="0"
        >
          13
        </li>
      </ul>
    </div>
    <div>
      Total:
       
      —
    </div>
  </div>
  <div
    class="ability"
  >
    <label
      id="cha-select-label"
    >
      Charisma
    </label>
    <div
      aria-labelledby="cha-select-label"
      class="container"
      data-testid="cha"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="button"
        data-testid="cha-button"
        type="button"
      >
        —
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <ul
        aria-activedescendant="—"
        class="list"
        data-testid="cha-list"
        role="listbox"
        style="display: none;"
        tabindex="-1"
      >
        <li
          aria-selected="true"
          class="item selected"
          data-testid="cha-data-blank"
          role="option"
          tabindex="0"
        >
          —
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="cha-data-10"
          role="option"
          tabindex="0"
        >
          10
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="cha-data-12"
          role="option"
          tabindex="0"
        >
          12
        </li>
        <li
          aria-selected="false"
          class="item"
          data-testid="cha-data-13"
          role="option"
          tabindex="0"
        >
          13
        </li>
      </ul>
    </div>
    <div>
      Total:
       
      —
    </div>
  </div>
</div>
`;
